File: src/main/java/com/yupi/airouter/controller/UserProviderKeyController.java
/**
 * 用户提供者密钥控制器（BYOK）
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
package com.yupi.airouter.controller;

import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.common.BaseResponse;
import com.yupi.airouter.common.DeleteRequest;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.exception.BusinessException;
import com.yupi.airouter.exception.ErrorCode;
import com.yupi.airouter.model.dto.byok.UserProviderKeyAddRequest;
import com.yupi.airouter.model.dto.byok.UserProviderKeyUpdateRequest;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.model.vo.UserProviderKeyVO;
import com.yupi.airouter.service.UserProviderKeyService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/byok")
public class UserProviderKeyController {

    @Resource
    private UserProviderKeyService userProviderKeyService;

    @Resource
    private UserService userService;

    /**
     * 添加用户提供者密钥
     */
    @PostMapping("/add")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "添加用户提供者密钥")
    public BaseResponse<Boolean> addUserProviderKey(@RequestBody UserProviderKeyAddRequest request,
                                                     HttpServletRequest httpRequest) {
        User loginUser = userService.getLoginUser(httpRequest);
        boolean result = userProviderKeyService.addUserProviderKey(request, loginUser.getId());
        return ResultUtils.success(result);
    }

    /**
     * 更新用户提供者密钥
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "更新用户提供者密钥")
    public BaseResponse<Boolean> updateUserProviderKey(@RequestBody UserProviderKeyUpdateRequest request,
                                                        HttpServletRequest httpRequest) {
        User loginUser = userService.getLoginUser(httpRequest);
        boolean result = userProviderKeyService.updateUserProviderKey(request, loginUser.getId());
        return ResultUtils.success(result);
    }

    /**
     * 删除用户提供者密钥
     */
    @PostMapping("/delete")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "删除用户提供者密钥")
    public BaseResponse<Boolean> deleteUserProviderKey(@RequestBody DeleteRequest request,
                                                        HttpServletRequest httpRequest) {
        if (request == null || request.getId() == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User loginUser = userService.getLoginUser(httpRequest);
        boolean result = userProviderKeyService.deleteUserProviderKey(request.getId(), loginUser.getId());
        return ResultUtils.success(result);
    }

    /**
     * 获取我的提供者密钥列表
     */
    @GetMapping("/my/list")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "获取我的提供者密钥列表")
    public BaseResponse<List<UserProviderKeyVO>> listMyProviderKeys(HttpServletRequest httpRequest) {
        User loginUser = userService.getLoginUser(httpRequest);
        List<UserProviderKeyVO> list = userProviderKeyService.listUserProviderKeys(loginUser.getId());
        return ResultUtils.success(list);
    }
}

File: go-backend/internal/controller/user_provider_key_controller.go
package controller

import (
	"log"

	"github.com/gin-gonic/gin"
	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type UserProviderKeyController struct {
	userProviderKeyService *service.UserProviderKeyService
	userService            *service.UserService
}

func NewUserProviderKeyController(
	userProviderKeyService *service.UserProviderKeyService,
	userService *service.UserService,
) *UserProviderKeyController {
	return &UserProviderKeyController{
		userProviderKeyService: userProviderKeyService,
		userService:            userService,
	}
}

func (c *UserProviderKeyController) AddUserProviderKey(ctx *gin.Context) {
	var request dto.UserProviderKeyAddRequest
	if err := ctx.ShouldBindJSON(&request); err != nil {
		log.Printf("add byok bind failed: %v", err)
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	result, serviceErr := c.userProviderKeyService.AddUserProviderKey(request, loginUser.ID)
	if serviceErr != nil {
		c.handleError(ctx, serviceErr)
		return
	}
	common.Success(ctx, result)
}

func (c *UserProviderKeyController) UpdateUserProviderKey(ctx *gin.Context) {
	var request dto.UserProviderKeyUpdateRequest
	if err := ctx.ShouldBindJSON(&request); err != nil {
		log.Printf("update byok bind failed: %v", err)
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	result, serviceErr := c.userProviderKeyService.UpdateUserProviderKey(request, loginUser.ID)
	if serviceErr != nil {
		c.handleError(ctx, serviceErr)
		return
	}
	common.Success(ctx, result)
}

func (c *UserProviderKeyController) DeleteUserProviderKey(ctx *gin.Context) {
	var request dto.DeleteRequest
	if err := ctx.ShouldBindJSON(&request); err != nil {
		log.Printf("delete byok bind failed: %v", err)
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	if request.ID == nil || request.ID.Int64() <= 0 {
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	result, serviceErr := c.userProviderKeyService.DeleteUserProviderKey(request.ID.Int64(), loginUser.ID)
	if serviceErr != nil {
		c.handleError(ctx, serviceErr)
		return
	}
	common.Success(ctx, result)
}

func (c *UserProviderKeyController) ListMyProviderKeys(ctx *gin.Context) {
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	list, serviceErr := c.userProviderKeyService.ListUserProviderKeys(loginUser.ID)
	if serviceErr != nil {
		c.handleError(ctx, serviceErr)
		return
	}
	common.Success(ctx, list)
}

func (c *UserProviderKeyController) handleError(ctx *gin.Context, err error) {
	if bizErr, ok := errno.AsBusinessError(err); ok {
		common.Error(ctx, bizErr.Code, bizErr.Message)
		return
	}
	common.Error(ctx, errno.SystemError.Code, "系统错误")
}
